pkg/billing/model.go

```diff
@@ -23,3 +26,44 @@ type IncrementalEvent struct {
 	StopTime       time.Time `json:"stop_time"`
 	Value          int       `json:"value"`
 }
+
+type enrichable interface {
+	setEventType()
+	idempotencyKey() *string
+}
+
+// Enrich sets the event's Type and IdempotencyKey fields, so that users of this API don't need to
+// manually set them
+//
+// Enrich is already called by (*Batch).Add, but this is used in the autoscaler-agent's billing
+// implementation, which manually calls Enrich in order to log the IdempotencyKey for each event.
+func Enrich[E enrichable](batch *Batch, event E) E {
```

Comment:
I think this doesn't need to be a generic, it's just a normal interface if you make the methods do the mutation themselves rather than needing to return  (I might, to avoid the whole "pointers everywhere" thing, have a "Init()` or `SetDefaults` method, that you can call in this or the batch interface.

Reply:
Updated.